### PR TITLE
Add search, sort and pagination feature to probes table

### LINKED ISSUE #65: Sort probes in UI based on execution order or name
Because of 
https://github.com/jenkins-infra/plugin-health-scoring/blob/6f9c5d2363df9599117e310f88f4461f61249ba5/src/main/java/io/jenkins/pluginhealth/scoring/http/ProbesController.java#L55

the probes are listed based on their name. However, it might be good to sort based on their execution order on each plugin. Or to have and indication of the execution order and to be able to sort the list as the user wants.

### PR DESCRIPTION
**Pull request by sridamul**
Closes #65 

It's been quite a long time since @dheerajodha and @lzeee tried to resolve this issue. This is a continuation of these PRs.

As mentioned by the @alecharp in this [comment](https://github.com/jenkins-infra/plugin-health-scoring/pull/222#issuecomment-1450151243), I tried to fetch the branch from @dheerajodha's repo which ultimately failed.
`git@github.com: Permission denied (publickey).`
Also the branch is too old to work on. So, I started working in a new branch (but major props to @dheerajodha and @lzeee for their contributions)

This PR replaces [datatables](https://datatables.net/) with [simple-datatables](https://github.com/fiduswriter/simple-datatables) which primarily uses plain JS instead of jquery.

Made changes according to the reviews given by @alecharp in the previous PR's. But still I have few doubts in them.

- [Comment-1](https://github.com/jenkins-infra/plugin-health-scoring/pull/109#discussion_r1022497645) 👇
> We have a package manager for front-end libraries and a build system to package everything.
This is not a working solution to me.

From my understanding and previous implementation by @lzeee , I removed the CDN of simple-database and installed the npm package of it, then abstracted the css into a separate file 'simple-datatable.less'. Do I need to change anything else?

- [Comment-2](https://github.com/jenkins-infra/plugin-health-scoring/pull/228#discussion_r1122805024) 👇
> This simple complex. Can we not have a <script> in the correct template and not have this in the global javascript?

As mentioned by @lzeee, we can use 'html-webpack-plugin' I have also researched about the alternatives but this is the most preferrable one. Suggestions are welcomed. also why don't we keep datatable function in the index.js?

Currently the probes page look like this:

https://github.com/jenkins-infra/plugin-health-scoring/assets/107404972/44e90838-deee-4aaa-a6d2-7c45d8e3d0f6

### CODE REVIEW

File: core/src/main/java/io/jenkins/pluginhealth/scoring/probes/DependabotProbe.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
 @Order(AbstractDependencyBotConfigurationProbe.ORDER)
 public class DependabotProbe extends AbstractDependencyBotConfigurationProbe {
     public static final String KEY = "dependabot";
+    public static final int ORDER = AbstractDependencyBotConfigurationProbe.ORDER;
```
Comment (alecharp) [inline core/src/main/java/io/jenkins/pluginhealth/scoring/probes/DependabotProbe.java:37]: In that case, make sure to use that `ORDER` in the `@Order` class annotation.

File: core/src/main/java/io/jenkins/pluginhealth/scoring/probes/RenovateProbe.java [hunk 2]
```diff
@@ -10,6 +10,7 @@
 @Order(AbstractDependencyBotConfigurationProbe.ORDER)
 public class RenovateProbe extends AbstractDependencyBotConfigurationProbe {
     public static final String KEY = "renovate";
+    public static final int ORDER = AbstractDependencyBotConfigurationProbe.ORDER;
```
Comment (alecharp) [inline core/src/main/java/io/jenkins/pluginhealth/scoring/probes/RenovateProbe.java:13]: Same.

**Comment by sridamul** [comment]
If possible, how can I add 'dheerajodha' and 'lzeee' as contributors to this PR?

**Review by alecharp — CHANGES_REQUESTED** [review]

**Comment by alecharp** [comment]
> If possible, how can I add 'dheerajodha' and 'lzeee' as contributors to this PR?

For this, you needed to add 

```
Co-Authored-By: Username <email>
```

in your commit messages. See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors.

It would be great to do that but it would mean to re-write your branch history and force push it. I'm usually against it but for this purpose, why not.

**Comment by sridamul** [comment]
> > If possible, how can I add 'dheerajodha' and 'lzeee' as contributors to this PR?
> 
> For this, you needed to add 
> 
> ```
> Co-Authored-By: Username <email>
> ```
> 
> in your commit messages. See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors.
> 
> It would be great to do that but it would mean to re-write your branch history and force push it. I'm usually against it but for this purpose, why not.

Right now I am moving back home, let me immediately do this after reaching home.

File: war/src/main/resources/templates/probes/listing.html [hunk 3]
```diff
@@ -2,6 +2,7 @@
 <html lang="en" data-layout-decorate="~{layouts/main}">
 <head>
     <title>Probes</title>
+    <script src="" data-th-src="@{/js/index.js}" type="text/javascript"></script>
```
Comment (alecharp) [inline war/src/main/resources/templates/probes/listing.html:5]: this is already included in the `layouts/main` which this page is based on.

```suggestion
```

File: war/src/main/js/index.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import {defineCustomElements} from "ionicons/dist/loader";
+import { DataTable } from "simple-datatables";
 
 defineCustomElements(window, {
   resourcesUrl: '/',
 });
 
+new DataTable("#probes-table");
```
Comment (alecharp) [inline war/src/main/js/index.js:8]: this means that the `DataTable` is created on all pages, even when there is no div with the correct id.
Would it be possible to only activate it on the `probes/listing.html`?

**Review by alecharp — CHANGES_REQUESTED** [review]

File: war/src/main/less/modules/simple-datatable.less
Comment (alecharp) [inline war/src/main/less/modules/simple-datatable.less:1]: I'm not sure why we need to have this entire file added in the project when it should come from the dependency directly.

Comment (sridamul) [reply war/src/main/less/modules/simple-datatable.less:1]: I thought it would be better to modify the design if we want in the future

Comment (alecharp) [reply war/src/main/less/modules/simple-datatable.less:1]: we shouldn't need the entire file but override the classes we want.

File: war/src/main/less/index.less [hunk 5]
```diff
@@ -7,6 +7,7 @@
 @import './modules/page-footer';
 @import './modules/score';
 @import './modules/table';
+@import '/node_modules/simple-datatables/src/css/style.css';
```
Comment (alecharp) [inline war/src/main/less/index.less:10]:
```suggestion
@import '~simple-datatables/src/css/style.css';
```

**Comment by sridamul** [comment]
Hey @alecharp, if possible, could you please check if any more changes are required?

**Comment by alecharp** [comment]
@sridamul please, do not force-push on pull request that had already been reviewed. This means I have to check everything again.

**Comment by alecharp** [comment]
@sridamul I'm sorry I pushed a few commits on your pull request. This is mostly to fix issues that are checked since Spotless setup and to finish the work you have done. 

Thank you for your contribution. I'm going to merge this shortly.

**Comment by sridamul** [comment]
Hey @alecharp Is it ready to ship?

**Comment by alecharp** [comment]
Currently no. There are plenty of conflicts.

**Comment by sridamul** [comment]
Hey @alecharp I have resolved the conflicts and the changes seems to work fine as expected.
Following are the new changes made:
- added getOrder() fn in CodeOwnershipProbe
- applied spotbugs (which is the reason to change the copyright year and alignment in some probes)

**Comment by alecharp** [comment]
> sorry for applying the spotless changes in single PR

that was required of you as you changed the probe codes. However, I'm not sure it was required to do so on the unit tests. I'm looking at it.

in the end, I like the sorting and search of the probes. But the Java code changes seems a bit too big for such a little feature. I'm not questioning the pull request, just the addition of the `getOrder` method. I'm wondering if just adding the UI feature without the order couldn't be enough.

**Comment by sridamul** [comment]
> I'm not questioning the pull request, just the addition of the getOrder method. I'm wondering if just adding the UI feature without the order couldn't be enough.

do you mean to drop the execution order column?

**Comment by alecharp** [comment]
> do you mean to drop the execution order column?

Yes.

**Comment by sridamul** [comment]
Sorry for the force push!

Here is the final table lookup...

final_table

wdyt?

File: war/package.json [hunk 6]
```diff
@@ -24,7 +24,8 @@
     "webpack-remove-empty-scripts": "1.0.4"
   },
   "dependencies": {
-    "ionicons": "7.3.1"
+    "ionicons": "7.3.1",
+    "simple-datatables": "^9.0.0"
```
Comment (alecharp) [inline war/package.json:28]:
```suggestion
    "simple-datatables": "9.0.0"
```

**Review by alecharp — CHANGES_REQUESTED** [review]
> Sorry for the force push!

That's fine.

That would be great.
Could I ask of you to make sure to have the `9.0.3` version of `simple-datatables` and not use the `^` (which introduce version uncertainty). Thanks.

**Review by alecharp — APPROVED** [review]
Thank you very much for your time and effort on this pull request.

**Pull request merged**